backends/p4fmt/attach.cpp

```diff
@@ -0,0 +1,1663 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "attach.h"
+
+#include <deque>
+#include <sstream>
+#include <string>
+
+#include "frontends/common/options.h"
+#include "frontends/p4/fromv1.0/v1model.h"
+#include "frontends/parsers/p4/p4parser.hpp"
+#include "ir/dump.h"
+
+namespace P4 {
+
+Visitor::profile_t Attach::init_apply(const IR::Node *node) {
+    LOG4("Program dump:" << std::endl << dumpToString(node));
+    listTerminators_init_apply_size = listTerminators.size();
+    vectorSeparator_init_apply_size = vectorSeparator.size();
+    return Inspector::init_apply(node);
+}
+
+void Attach::end_apply(const IR::Node *) {
+    if (outStream != nullptr) {
+        cstring result = builder.toString();
+        *outStream << result.c_str();
+        outStream->flush();
+    }
+    BUG_CHECK(listTerminators.size() == listTerminators_init_apply_size,
+              "inconsistent listTerminators");
+    BUG_CHECK(vectorSeparator.size() == vectorSeparator_init_apply_size,
+              "inconsistent vectorSeparator");
+}
+
+// Try to guess whether a file is a "system" file
+bool Attach::isSystemFile(cstring file) {
+    if (noIncludes) return false;
+    if (file.startsWith(p4includePath)) return true;
+    return false;
+}
+
+cstring Attach::ifSystemFile(const IR::Node *node) {
+    if (!node->srcInfo.isValid()) return nullptr;
+    auto sourceFile = node->srcInfo.getSourceFile();
+    if (isSystemFile(sourceFile)) return sourceFile;
+    return nullptr;
+}
+
+namespace {
+class DumpIR : public Inspector {
+    unsigned depth;
+    std::stringstream str;
```

Comment:
Since you're doing *lots* of string concatenations, it is much better to use `Cord` from Abseil library.